controllers/local_db_statefulset.go

```diff
@@ -219,9 +220,12 @@ func (r *BackstageReconciler) patchLocalDbStatefulSetObj(statefulSet *appsv1.Sta
 
 func (r *BackstageReconciler) setDefaultStatefulSetImage(statefulSet *appsv1.StatefulSet) {
 	visitContainers(&statefulSet.Spec.Template, func(container *v1.Container) {
-		if len(container.Image) == 0 || container.Image == fmt.Sprintf("<%s>", bs.EnvPostGresImage) {
-			container.Image = r.PsqlImage
+		if val, ok := os.LookupEnv(EnvPostgresImage); ok {
```

Comment:
This system call get invoked during each reconciliation and can be avoided if the operator reads the env vars during the startup and use the values later.

Reply:
True, and having it reconsiled we can change configuration dynamically, not requiring restarting controller and interrupting service, which is a Kubernetes way I believe.

Reply:
What do you mean by "change configuration dynamically"? I think the issue here is reading the `RELATED_IMAGE_*` env vars from the system during each reconciliation, which is separate from the configuration.
If any of the env vars are modified, K8s will automatically restart the controller pod, I think. In this case, I agree with @jianrongzhang89 that it would probably make more sense to read the env vars only once at startup to initialize the controller..

Reply:
I think I do not understand why for simple reading env variable we need to make additional code and pull it through the whole application... What for?

Reply:
You mentioned that "having it reconsiled we can change configuration dynamically, not requiring restarting controller and interrupting service", but in the case of env vars, we won't be able to change them dynamically without a controller restart.
So instead of reading the same env var (and getting the same value) during each reconciliation (and for all containers), we could avoid that by setting the values at initialization time (as done previously).

Reply:
In that case, maybe you could read the value before visiting all containers, rather than reading it in every container visitor func.

```go
  if val, ok := os.LookupEnv(EnvPostgresImage); ok {
	  visitContainers(&statefulSet.Spec.Template, func(container *v1.Container) {
		  container.Image = val
	  })
  }
```

Reply:
ok

Reply:
> I think I do not understand why for simple reading env variable we need to make additional code and pull it through the whole application... What for?

This is for code quality. The repeated function calls should be avoided whenever it is possible. This can be easily achieved by storing the data in a global variable.

Reply:
> True, and having it reconsiled we can change configuration dynamically, not requiring restarting controller and interrupting service, which is a Kubernetes way I believe.

I do not think it supports dynamic configuration change. os.LookupEnv() won't pick up new env variable values unless the pod gets restarted.

Reply:
Not sure I understand your point about "repeated function". Could you please show me the place (this code is outdated already)